### ISSUE TITLE
Kubelet does not report NodeReady status correctly

### ISSUE THREAD

**Issue by davidopp** [GoogleCloudPlatform/kubernetes#8585]
This code in kubelet.go is clearly wrong (I noticed this when I was reviewing #7170; see my comment in this section of code there for more details). Whether it's related to the problems we're seeing, I'm not sure. But either way this is surely something we need to fix for 1.0. 

```
	updated := false
	for i := range node.Status.Conditions {
		if node.Status.Conditions[i].Type == api.NodeReady {
			newCondition.LastTransitionTime = node.Status.Conditions[i].LastTransitionTime
			if node.Status.Conditions[i].Status != api.ConditionTrue {
				kl.recordNodeOnlineEvent()
			}
			node.Status.Conditions[i] = newCondition
			updated = true
		}
	}
	if !updated {
		newCondition.LastTransitionTime = currentTime
		node.Status.Conditions = append(node.Status.Conditions, newCondition)
		kl.recordNodeOnlineEvent()
	}
```

@dchen1107 @lavalamp @fgrzadkowski

**Comment by davidopp** [comment]
I just noticed that @simon3z pushed a new version of his PR in response to my comment on this section; I haven't checked to see whether his change fixes this problem.

**Comment by davidopp** [comment]
I don't think his change fixes it. I think the logic we want is something like this (I'm running off to a meeting, so this is probably not completely correct)

```
if (oldReadyNodeConditionStatus == newReadyNodeCondition.Status) {
   newReadyNodeCondition.LastTransitionTime = node.Status.Conditions[i].LastTransitionTime
}
else {
  newReadyNodeCondition.LastTransitionTime = currentTime
  // also generate appropriate event
}
newReadyNodeCondition.LastUpdateTime = currentTime
node.Status.Conditions[i] = newReadyNodeCondition
```

**Comment by yujuhong** [comment]
kubelet used to post only ready status until #7763 was merged. The logic below was not modified to account for this change.

**Comment by simon3z** [comment]
@davidopp I am trying to handle this in #7170

**Comment by davidopp** [comment]
Yep, can you please add "Fixes #8585" to the commit message for that PR?
